crates/bevy_hierarchy/src/hierarchy.rs

```diff
@@ -198,6 +202,50 @@ impl<'w> DespawnRecursiveExt for EntityWorldMut<'w> {
     }
 }
 
+/// Trait that holds functions for cloning entities recursively down the hierarchy
+pub trait CloneEntityRecursiveExt {
+    /// Sets the option to recursively clone entities.
+    /// When set to true all children will be cloned with the same options as the parent.
+    fn recursive(&mut self, recursive: bool) -> &mut EntityCloneBuilder;
+}
+
+impl CloneEntityRecursiveExt for EntityCloneBuilder {
+    fn recursive(&mut self, recursive: bool) -> &mut EntityCloneBuilder {
+        if recursive {
+            self.override_component_clone_handler::<Children>(ComponentCloneHandler::Custom(
+                component_clone_children,
+            ))
+            .override_component_clone_handler::<Parent>(ComponentCloneHandler::Ignore)
+        } else {
+            self.override_component_clone_handler::<Children>(ComponentCloneHandler::Default)
+                .override_component_clone_handler::<Parent>(ComponentCloneHandler::Default)
+        }
+    }
+}
+
+/// Clone handler for the [`Children`] component. Allows to clone the entity recursively.
+fn component_clone_children(
+    world: &mut World,
+    _component_id: ComponentId,
+    entity_cloner: &EntityCloner,
+) {
+    let children = world
+        .get::<Children>(entity_cloner.get_source())
+        .expect("Source entity must have Children component")
```

Comment:
Do we need to panic here? Cloning children with no children could just be a no-op.

Reply:
This function is going to be called only for `Children` component, if it was called for some other component (like if someone manually overridden it) it is definitely an error. The function isn't actually public, so user can't set it as a handler anyway, so I though panicking in this context is the right call.
But generally we can just log it as an error instead, It's just that this kind of situation is probably not going to happen unless someone is intentionally doing something wrong.